Config plugins

a4z opened the pull request:
this is a suggestion to solve #118 through allowing to specify optional files that can be used to add and configure plugins.
Please see changes in README.adoc for more information

a4z commented:
no idea why or what on travis if failing. changes are working great here. please help!

obilodeau commented:
Thanks for your contribution! Sorry I haven't replied earlier. A quick glance and everything looks fine. Plus, I like the design. I'm going to test it later and report back.

Travis failures are related to the fact that your code changed the output of the generated HTML. We do strict regression testing against these. Don't worry about it, I can fix that.

One thing I'm wondering: I'm not sure attributes is a list, I think it's a hashmap (key / value) so multiple  `revealjs_plugins_add` statements will overwrite each other. I'll verify that by trying to add two plugins.

a4z commented:
Thanks for your feedback!
multiple revealjs_plugins_add makes no sense, the file file given needs to have a comma separated list of file path to plugins

example from what I use:

cat plugins_add.js
```
{ src: 'config/plugins/mouse-pointer.js' },
{ src: 'config/plugins/badges.js' }
```

cat plugins_setup.js
```
       badges: {
          languages: true
        }
```

if the concept of this solution will be approved I can improve the documentation and put up a sample 


knowing that it is just about some (maybe withe spaces) in a test seems fixable, but I am on holidays this week, can have a look at it in 2 weeks if you have not time

obilodeau commented on templates/document.html.slim:
```diff
@@ -21,7 +21,7 @@ html lang=(attr :lang, 'en' unless attr? :nolang)
     - if attr? :icons, 'font'
       - if attr? 'iconfont-remote'
-        link rel='stylesheet' href=(attr 'iconfont-cdn', %(#{cdn_base}/font-awesome/4.3.0/css/font-awesome.min.css))
+        link rel='stylesheet' href=(attr 'iconfont-cdn', %(#{cdn_base}/font-awesome/4.5.0/css/font-awesome.min.css))
```
This should be done in a separate PR

a4z commented:
I am glad you like the solution! 
Thanks for you effort to fix the open points and your work to pull this in!

obilodeau commented:
Made the discussed changes. Feel free to comment. Will merge if no objections.

Thanks for your contribution @a4z!

a4z commented:
Glad you like it!
The only thing I noticed since last time, and where I started to think about how to solve it, is that the notes plugin (where you see the next slide in a new window) does not work via revealjs over CDN.
In this case the node plugin needs to be added manual to a custom folder.
So maybe it should be removed from the default loaded plugins.
Also, the loaded markdown plugins may not be used ever, and loading them should be avoided.
So I start to think if there should be any plugin loaded per default....

obilodeau commented:
I use speaker notes extensively and I never noticed. I never used reveal.js hosted by a CDN because I don't trust conference networks :wink:. It's probably blocked because of a same origin policy violation. I know speaker notes relies on PostMessage to pass state around. Are you sure hosting only the notes plugin locally fixes the issue?

Also, if you want reveal.js hosted on a CDN, isn't hosting plugins locally defeating the purpose? 

I think the issues you raise are interesting but I would file a separate issue for them and merge this PR. What do you think?

a4z commented:
I use the CDN especially during creation of presentation but also for talks company intern since I switch computers very often, and this makes things  easy. so far I never encountered any problem, but for important talks I have everything as a bundle.
yes, copying the plugin to a user defined folder. like custom plugins, and load it from there work with revealjs via CDN. Just not load it twice, with the default path, and a custom one.
That's why I had to change the default template, and when I was there I also removed the plugins I do not need because things load slightly faster if this is omitted.

I am thinking about striping out the whole plugin configuration section into a file, and load it, since I have always a config file. And most of the default loaded plugins are not used by me so I see no point in loading them, while I add other.
I fixed this 10 minute before a talk in the company I am working, and did not have time to thing about since then.
And currently I am on vacation, so I think I will not have something this week,

but the first thought is, why not strip out the whole section if there is something custom give, otherwise use the default ....
it should theoretical be possible to read an other file if no file is given, so ... 
will try it after vacation

a4z commented:
Thanks for your effort and improvements @obilodeau !!
